### DISCUSSION TITLE
Questions about the promise modules protocol

### DISCUSSION THREAD

**Discussion by amousset** [cfengine/core#4481] [Q&A] [answered]
Hi everyone!

I'd like to thank you for the addition of the new promise modules that allow powerful extension of CFEngine!

I'm starting to implement some, and have a couple of questions about the modules and protocol specs.

### result_classes

In the python implementation I read:

> evaluate_promise should return either a result or a (result, result_classes) pair

But don't see that mentioned in the specs (only the result field). Is it part of the protocol?

### action_policy

In the specs I read:

> action_policy, if not default, will be sent to the module, for dry-run/no-changes functionality

But when trying to use an action body I get:

```
./main.cf:20:0: error: Common attribute 'action' not implemented for custom promises (git)
```

### promise_type parameter

For validation operation in JSON, the spec example gives:

``` json
{"operation": "evaluate_promise", "log_level": "info", "promise_type": "git", "promiser": "/opt/cfengine/masterfiles", "attributes": {"repo": "https://github.com/cfengine/masterfiles"}}
```

But when testing, it appears the `promise_type` parameter is not passed.

### Using a binary module

The interpreter parameter is mandatory, which makes using a binary promise module a bit weird. I managed to run it with:

```cfengine
promise agent git
{
  interpreter => "/path/to/module";
  # We ignore it in the module
  path => "/path/to/module";
}
```

**Comment by amousset** [comment]
Another one:

```
{"operation": "terminate", "log_level": "info"}
```
in the doc but `log_level` does not seem to be there.

**Reply by amousset** [reply]
FYI I'm currently working on a Rust implementation in https://github.com/Normation/cfengine-promise-rust

**Reply by olehermanse** [reply]
> FYI I'm currently working on a Rust implementation in https://github.com/Normation/cfengine-promise-rust

Very cool! Be sure to update us on releases :)

**Reply by olehermanse** [reply]
> in the doc but log_level does not seem to be there.

Thanks, created a ticket for this as well: https://tracker.mender.io/browse/CFE-3564

**Comment by olehermanse** [comment, marked as answer]
Hey, @amousset!

Thanks for testing out promise modules, a complete overview of what is currently implemented and what is missing can be found in the epic:

https://tracker.mender.io/browse/CFE-1939

> In the python implementation I read:
>
> evaluate_promise should return either a result or a (result, result_classes) pair
>
> But don't see that mentioned in the specs (only the result field). Is it part of the protocol?

Thank you for pointing this out. I have updated the spec here:

https://github.com/cfengine/core/pull/4482

> Common attribute 'action' not implemented for custom promises (git)

Correct, ticket to implement is here: https://tracker.mender.io/browse/CFE-3433

> But when testing, it appears the promise_type parameter is not passed.

Created a bug ticket to fix this here: https://tracker.mender.io/browse/CFE-3563

> The interpreter parameter is mandatory, which makes using a binary promise module a bit weird.

Created a ticket to improve this: https://tracker.mender.io/browse/CFE-3562

Hope that helps!
